[WIP][5.6] Use notifications as event listeners

rasmuscnielsen opened the pull request:
****Proposal****

This PR is a first draft of implementing Notifications as Event Listeners.

Given you have the following event 

```php
class TeamMemberAdded
{
    public $team, $teamMember;

    // ... constructor...
}
```

and the following notification

```php
class TeamMemberWelcomeNotification extends Notification
{
    public function routeNotificationForEvent($event)
    {
        return $event->teamMember;
    }

    public function __construct($team)
    {
        $this->team = $team;
    }
} 
```
... you could subscribe to it just by adding it as a listener.

```php
Event::listen(TeamMemberAdded::class, TeamMemberWelcomeNotification::class);
```

This PR adresses two issues on using notifications as event listeners

**1. It adds a `@handle` on `Notification` class that dispatches the notification to the right notifiable**

This part is heavily Inspired by @themsaid 's post https://themsaid.com/laravel-mailables-notfiications-as-event-listeners-20170516 

The issue I found with this approach was however, that it made it inconvenient to send the notification normally, because one has to remove any constructor-arguments and then set them manually.
Which leads to the next issue:

**2. It makes the Event Dispatcher instantiate listeners with the public properties on the event**

Given the event fired is an object and the listener is a qualified class, it now attempts to construct the listener with the public properties of the class.
This is what makes the above example inject '$team' into the notification constructor.

However this introduces a potential compatibility issue. If somebody is already type hinting a `$team` variable, it would now break:

```php
class SomeOtherListener
{
    public function __construct(SomeOtherTeam $team)
    {
        $this->team = $team;
    }
} 
```  

Any suggestions on the above issue is appreciated.

**UPDATE 2017-11-05**

Added a ReflectOnEvent interface that tells the EventDispatcher that this listener would like to be resolved with the public properties of the event. 
I feel this is a quite clean way to opt-in on the new behavior.

**TODO**

- Make it compatible with queueable listeners

Dylan-DPC commented:
This is a breaking change. you are modifying signatures of existing functions

Dylan-DPC commented:
I am not sure have to look more closely.

Dylan-DPC commented:
Haven't checked it closely yet. You need to rebase your branch to remove the unnecessary commits

rasmuscnielsen commented:
@Dylan-DPC Thx I changed it [5.6] and changed base

Any thoughts on the feature itself?

To avoid it becoming an issue when upgrading, I was thinking about introducing an `ReflectOnEvent` interface that the `Notification` base class would then implement. This way the reflection is easily portable for other listeners, but no forced on existing ones.

rasmuscnielsen commented:
@Dylan-DPC Thx I will dod that - thought I'd wait till the feature was fully implemented though? 

Still have this TODO (from the PR description)

- Find a way to ensure backwards compatibility (note: see my suggestion in last comment)
- Make it compatible with queueable listeners

sisve commented:
I think you should add the [WIP] tag to the pr title since this is not yet done and there's a TODO section in your description.

Dylan-DPC commented:
> Find a way to ensure backwards compatibility (note: see my suggestion in last comment)

@rasmuscnielsen you don't need to worry about backwards compatibility if this is going to be implemented. It will be good if it is BC but shouldn't be a priority

rasmuscnielsen committed:
Add support for event listening for sync notifications

rasmuscnielsen commented:
@Dylan-DPC I see what you meant on the rebasing - it was due I originally checked it out from v5.5 and afterwards changed brands. Fixed it.

Regarding backward compatibility I feel it's weird upgrade to say that 'your event listeners might break if there is a naming collision with the public properties of the event'.

I'd love to hear your thoughts when you found the time to check it out. I will try to implement the `ReflectOnEvent` interface later which I think could be a good solution.

rasmuscnielsen committed:
Introduce ReflectOnEvent for notifications

rasmuscnielsen commented:
Added a ReflectOnEvent interface that tells the EventDispatcher that this listener would like to be resolved with the public properties of the event. 
I feel this is a quite clean way to opt-in on the new behavior.

Regarding the queueable listeners I think it would be nice with some indication whether this proposal is well-received before spending a lot of time digging into the rabbit hole of queued jobs :-)

taylorotwell commented:
I don't understand all this "reflectsOnEvent" stuff. Seems weird and confusing.

themsaid commented:
@rasmuscnielsen Honestly I'm not very excited about this, I think it's adding some complexity to the current code with not too much value in return, a notification is a `Notification` it's not a `Listener`, if you want to use it as a listener you can already do that using the workaround in my blog without having to do any changes in the code base, and you can easily overcome the downside of wanting to send the notification not as listener by having your constructor arguments as optional and manage the dependencies internally however you want.

I'd like to add that I encourage this approach ONLY for when the use case is pretty simple, but I I were you and I have a notification that could be send inside and event or manually, I'll just create a listener class that sends the notification, no need to add changes to the core just for this simple task.

taylorotwell commented:
I don't think I will be adding this implementation, though I appreciate the sentiment of wanting to get rid of the class that simply fires a notification.

rasmuscnielsen commented:
@taylorotwell @themsaid Thank you both very much for taking the time to review this PR. I appreciate the feedback!

For anyone interested, I ended up with a simpler solution that solves the same problem.

Instead of changing the behavior of the dispatcher, instead I introduced a decorator.

```php 
Event::listen(ChannelAuthorized::class, new EventListener(UpdateChannelResources::class)); // decorating a job
``

```php
<?php

namespace App;

use Illuminate\Contracts\Notifications\Dispatcher;
use Illuminate\Notifications\Notification;

class EventListener
{
    /**
     * @var
     */
    protected $listenerClass;

    /**
     * EventListener constructor.
     * @param $listenerClass
     */
    public function __construct($listenerClass)
    {
        $this->listenerClass = $listenerClass;
    }

    /**
     * @param $event
     * @return mixed
     */
    public function __invoke($event)
    {
        $listener = app()->makeWith($this->listenerClass, get_object_vars($event));

        if ($this->isDispatchable($listener)) {
            dispatch($listener);
        }
        elseif ($this->isNotification($listener)) {
            app(Dispatcher::class)->send($listener->routeNotificationForEvent($event), $listener);
        }
        else {
            $listener->handle($event);
        }
    }

    /**
     * @param $listener
     * @return bool
     */
    protected function isDispatchable($listener)
    {
        return method_exists($listener, 'dispatch');
    }

    /**
     * @param $listener
     * @return bool
     */
    protected function isNotification($listener)
    {
        return $listener instanceof Notification;
    }
}
```
